File: src/Json/Process/FactorValue.fs
namespace ARCtrl.Json


open Thoth.Json.Core

open ARCtrl
open ARCtrl.Process
open System.IO


module FactorValue =
    
    let genID (fv : FactorValue) = 
        match fv.Category,fv.Value,fv.Unit with
        | Some t, Some v, Some u -> $"#Factor_{t.NameText}={v.Text}{u.NameText}"
        | Some t, Some v, None-> $"#Factor_{t.NameText}={v.Text}"
        | _ -> $"#EmptyFactor"

    module ISAJson = 
        
        let encoder (idMap : IDTable.IDTableWrite option) (fv : FactorValue) = 
            let f (fv : FactorValue) =
                [
                    // Is this required for ISA-JSON? The FactorValue type has an @id field
                    Encode.tryInclude "@id" Encode.string (fv |> genID |> Some)
                    Encode.tryInclude "category" (Factor.ISAJson.encoder idMap) fv.Category
                    Encode.tryInclude "value" (Value.ISAJson.encoder idMap) fv.Value
                    Encode.tryInclude "unit" (OntologyAnnotation.ISAJson.encoder idMap) fv.Unit
                ]
                |> Encode.choose
                |> Encode.object
            match idMap with
            | None -> f fv
            | Some idMap -> IDTable.encode genID f fv idMap

        let decoder: Decoder<FactorValue> =
            Decode.object (fun get ->
                {
                    ID = get.Optional.Field "@id" Decode.uri
                    Category = get.Optional.Field "category" Factor.ISAJson.decoder
                    Value = get.Optional.Field "value" Value.ISAJson.decoder
                    Unit = get.Optional.Field "unit" OntologyAnnotation.ISAJson.decoder
                }
            )

File: js/Json/Process/FactorValue.js
import { genID, decoder, encoder } from "../PropertyValue.js";
import { FactorValue, FactorValue_createAsPV } from "../../Core/Process/FactorValue.js";
import { ofArray, choose } from "../../fable_modules/fable-library-js.4.22.0/List.js";
import { map } from "../../fable_modules/fable-library-js.4.22.0/Option.js";
import { tryInclude } from "../Encode.js";
import { decoder as decoder_1, encoder as encoder_1 } from "./Factor.js";
import { decoder as decoder_2, encoder as encoder_2 } from "./Value.js";
import { OntologyAnnotation_ISAJson_decoder, OntologyAnnotation_ISAJson_encoder } from "../OntologyAnnotation.js";
import { map as map_1 } from "../../fable_modules/fable-library-js.4.22.0/Seq.js";
import { encode } from "../IDTable.js";
import { object } from "../../fable_modules/Thoth.Json.Core.0.4.0/Decode.fs.js";
import { Decode_uri } from "../Decode.js";

export const ROCrate_encoder = encoder;

export const ROCrate_decoder = decoder(FactorValue_createAsPV);

export function ISAJson_genID(fv) {
    return genID(fv);
}

export function ISAJson_encoder(idMap, fv) {
    const f = (fv_1) => {
        const values = choose((tupledArg) => map((v_1) => [tupledArg[0], v_1], tupledArg[1]), ofArray([tryInclude("@id", (value) => ({
            Encode(helpers) {
                return helpers.encodeString(value);
            },
        }), ISAJson_genID(fv_1)), tryInclude("category", (value_2) => encoder_1(idMap, value_2), fv_1.Category), tryInclude("value", (value_3) => encoder_2(idMap, value_3), fv_1.Value), tryInclude("unit", (oa) => OntologyAnnotation_ISAJson_encoder(idMap, oa), fv_1.Unit)]));
        return {
            Encode(helpers_1) {
                const arg = map_1((tupledArg_1) => [tupledArg_1[0], tupledArg_1[1].Encode(helpers_1)], values);
                return helpers_1.encodeObject(arg);
            },
        };
    };
    if (idMap != null) {
        return encode(ISAJson_genID, f, fv, idMap);
    }
    else {
        return f(fv);
    }
}

export const ISAJson_decoder = object((get$) => {
    let objectArg, objectArg_1, objectArg_2, objectArg_3;
    return new FactorValue((objectArg = get$.Optional, objectArg.Field("@id", Decode_uri)), (objectArg_1 = get$.Optional, objectArg_1.Field("category", decoder_1)), (objectArg_2 = get$.Optional, objectArg_2.Field("value", decoder_2)), (objectArg_3 = get$.Optional, objectArg_3.Field("unit", OntologyAnnotation_ISAJson_decoder)));
});

